docs/design/rosbag2_playback_time.md

```diff
@@ -0,0 +1,115 @@
+
+```
+class PlayerClock : public rclcpp::Clock
+{
+public:
+  /**
+   * playback_rate must be positive, 1.0 means real-time
+   * starting_time provides an offset for all future times
+   * clock_topic_publish_frequency:
+   *   - if > 0 - publishes `now()` to /clock at the specified frequency in Hz, rate determined by a system steady clock
+   *   - if <= 0 - does not publish the /clock topic
+   */
+  Clock(float playback_rate, rclcpp::Time starting_time, float clock_topic_publish_frequency);
+
+  // rclcpp::Clock interface
+  /**
+   * based on an internal steady clock, calculates current "Player Time"
+   */
+  rclcpp::Time now() override;
+
+  /**
+   * Sleeps (non-busy-waiting) the current thread until the provided time is reached - according to this Clock!
+   *
+   * If time is paused, the requested time may never be reached
+   * `real_time_timeout` uses the internal steady clock, if the timeout elapses, return false
+   * If jump() is called, return false, allowing the caller to handle the new time
+   * Return true if the time has been reached
+   */
+  bool sleep_until(rclcpp::Time until, rclcpp::Duration real_time_timeout) override;
+
+  // Special playback interface
+
+  /**
+   * Pauses/resumes time.
+   * While paused, `now()` will repeatedly return the same time, until resumed
+   *
+   * Note: this could have been implemented as `set_rate(0)`, but this interface allows this clock to maintain the
+   * clock's rate internally, so that the caller does not have to save it in order to resume.
+   */
+  void set_paused(bool paused);
+
+  /**
+    * Changes the rate of playback. Must be positive (nonzero)
+    */
+  void set_rate(float rate);
+
+  /**
+    * Changes the current internally maintained offset so that next published time is different.
+    * This should trigger any JumpHandler callbacks that are registered with the clock (see "Time Jumps" section)
+    */
+  void jump(rclcpp::Time time);
+
+}
+```
+
+### Pause and Resume
+
+No special allowance needs to be made in the `Player` - time will still be provided, but it will not move forward, so the next message will not be published until time is resumed.
+
+### Setting the Rate
+
+No special allowance needs to be made in the `Player` - time will be provided at a different rate, and `sleep_until` will handle this.
+
+### Time Jumps
+
+The `Player` must register a `JumpHandler` with the `Clock` - so that when a jump occurs, the current message playback queue can be invalidated and re-enqueued according to the new starting time.
+
+
+## Open questions
+
+* Should time interpolation in the ROS clock be used for external time driver? If the publish rate is too slow, playback cannot reach any close fidelity if there is none. E.g. a topic publishing at 200Hz would need at least a 200Hz `/clock`, unless the `Clock` can interpolate between samples.
```

Comment:
IIUC, I don't think we can reliably ensure that time interpolation will always work for external time drivers.
If this is the case, rosbag2 can either provide the 'best effort' rate and/or inform the user via some warning it was unable to publish at the requested rate.

Reply:
I was thinking we could do something simple but reasonably effective to acieve interpolation on the receiving end, based on e.g. the last 2 samples plus a deadline timeout to handle when no new clock is received.

Reply:
Without any interpolation, e.g. "discrete time" - faster messages would still be published in the right order and at the right "rate", but they'd be bursty, being published in groups immediately after receiving each clock sample, rather than smoothly publishing through the gaps between clock samples. e.g.

```
# dots are arbitrary time tick, letters are when a message is published
# M is topic Message, C is /clock message

# real time (or with interpolation
M..M..M..M..M..M..M..M..M..M..M..
C.......C........C........C......

# without any interpolation if clock is slow
M.......MM.......MMM......MMM....
C.......C........C........C......
```